File: 8-string-to-integer-atoi/8-string-to-integer-atoi.py
import re

class Solution:
    def myAtoi(self, s: str) -> int:
        maxValue = 2 ** 31 - 1
        minValue = -2 ** 31
        
        i = 0
        ans = 0
        sign = 1
        
        # whitespace
        while i < len(s) and s[i] == ' ':
            i += 1
            
        # +/-
        if i < len(s) and s[i] == '-':
            i += 1
            sign = -1
        elif i < len(s) and s[i] == '+':
            i += 1
            
        # check number 0-9
        digits = set('0123456789')
        while i < len(s) and s[i] in digits:
            ans = ans * 10 + int(s[i])
            i += 1
            
        # check bounds
        ans *= sign
        
        if ans < 0:
            return max(ans, minValue)
        return min(ans, maxValue)
        

File: 8-string-to-integer-atoi/8-string-to-integer-atoi.rb
# @param {String} s
# @return {Integer}
def my_atoi(s)
    max_value = 2 ** 31 - 1
    min_value = -2 ** 31
    
    i, ans, sign = 0, 0, 1
    
    # whitespace
    while i < s.length && s[i] == ' '
        i += 1
    end
    
    # +/-
    if i < s.length && s[i] == '-'
        sign = -1
        i += 1
    elsif i < s.length && s[i] == '+'
        i += 1
    end
    
    # check digits
    digits = '0123456789'
    while i < s.length && digits.include?(s[i])
        ans = ans * 10 + s[i].to_i
        i += 1
    end
    
    ans *= sign
    
    return ans < 0 ? [ans, min_value].max : [ans, max_value].min
end